shell: extract from daemon and reuse it for Cilium Operator

mhofstetter opened the pull request:
This PR extracts the Cilium debug shell cell from the daemon and reuses it for the Cilium Operator. In addition it adds the `shell` command (to connect to the debug shell)  to the Cilium Agent & Operator binaries.

Although there are currently not many shell commands available for the Cilium operator (more or less only accessing statedb tables - with health being the only table), this could serve as a starting point for using the shell to check the operator runtime (by adding more shell commands in the future).

**Example (with local kind cluster)**

Operator

```
❯ kubectl -n kube-system exec -it cilium-operator-86649d6bc8-sfp9v -- cilium-operator-generic shell
time="2025-04-07T13:18:09.263802882Z" level=debug msg="Skipped reading configuration file" error="Config File \"cilium-operators\" Not Found in \"[/]\"" subsys=config
    /¯¯\
 /¯¯\__/¯¯\
 \__/¯¯\__/  Cilium 1.18.0-dev d4f2da37d4 2025-04-07T14:28:01+02:00 go version go1.24.2 linux/amd64
 /¯¯\__/¯¯\  Welcome to the Cilium Shell! Type 'help' for list of commands.
 \__/¯¯\__/
    \__/

kind-worker>
```

Agent

```
❯ kubectl -n kube-system exec -it cilium-hgnpr  -- cilium-agent shell
time="2025-04-07T13:18:24.45746434Z" level=info msg="Memory available for map entries (0.250% of 32338743296B): 80846858B" subsys=config
time="2025-04-07T13:18:24.457598437Z" level=info msg="option bpf-ct-global-tcp-max set by dynamic sizing to 283673" subsys=config
time="2025-04-07T13:18:24.457608005Z" level=info msg="option bpf-ct-global-any-max set by dynamic sizing to 141836" subsys=config
time="2025-04-07T13:18:24.45761904Z" level=info msg="option bpf-nat-global-max set by dynamic sizing to 283673" subsys=config
time="2025-04-07T13:18:24.457631333Z" level=info msg="option bpf-neigh-global-max set by dynamic sizing to 283673" subsys=config
time="2025-04-07T13:18:24.457644323Z" level=info msg="option bpf-sock-rev-map-max set by dynamic sizing to 141836" subsys=config
time="2025-04-07T13:18:24.457753137Z" level=warning msg="Running Cilium with \"kvstore\"=\"\" requires identity allocation via CRDs. Changing identity-allocation-mode to \"crd\"" subsys=config
    /¯¯\
 /¯¯\__/¯¯\
 \__/¯¯\__/  Cilium 1.18.0-dev d4f2da37d4 2025-04-07T14:28:01+02:00 go version go1.24.2 linux/amd64
 /¯¯\__/¯¯\  Welcome to the Cilium Shell! Type 'help' for list of commands.
 \__/¯¯\__/
    \__/

kind-control-plane>
```

Unfortunately depending on the config triggers some log messages - depending on the log level (`debug` in above case). But i think these are fine for the moment - and should be eventually eliminated by further reducing the global config.

Note:

- It still uses a UNIX domain socket as shell socket -> I thought about using a tcp socket on the loopback interface and accessing it via `kubectl port-forward` functionality. This way we could implement the shell client functionality within the Cilium CLI.
- This PR also fixes that stopped shell sessions are listed in the health table (by properly closing the health scope).
- Currently one needs to know the name of the binary (e.g. `cilium-operator-generic`)

mhofstetter committed:
daemon: extract shell into pkg/shell

This commit moves the shell cell from the package
`daemon/cmd` into `pkg/shell/server`. This way it can also be
used by Cilium components other then the Agent.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
operator: add shell

This commit adds the Cilium debug shell to the Cilium Operator
deployment.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
shell: close health scope for stopped shell sessions

Currently, ended shell sessions (that are internally handled
as hive jobs) are still listed in the health state db table.

```
kind-worker> db/show health
Module                                                                     Component                          Level     Message                                                 Error   LastOK                 UpdatedAt              Count
health                                                                     job-module-status-metrics          OK        Running                                                         2025-04-04T09:02:43Z   2025-04-04T09:02:43Z   1
operator.operator-controlplane.leader-lifecycle.controller-runtime         job-manager                        OK        Running                                                         2025-04-04T09:02:44Z   2025-04-04T09:02:44Z   1
operator.operator-controlplane.leader-lifecycle.enabled-features           job-update-config-metric           Stopped   Updating metrics based on OperatorConfig                        2025-04-04T09:02:45Z   2025-04-04T09:02:45Z   1
operator.operator-controlplane.leader-lifecycle.lbipam                     job-lbipam-main                    OK        Running                                                         2025-04-04T09:02:44Z   2025-04-04T09:02:44Z   1
operator.operator-controlplane.leader-lifecycle.network-policy-validator   observer-job-ccnp-validation       OK        Primed                                                          2025-04-04T09:02:45Z   2025-04-04T09:02:45Z   1
operator.operator-controlplane.leader-lifecycle.network-policy-validator   observer-job-cnp-validation        OK        Primed                                                          2025-04-04T09:02:45Z   2025-04-04T09:02:45Z   1
operator.operator-controlplane.leader-lifecycle.service-resolver           job-service-reloader-initializer   OK        Running                                                         2025-04-04T09:02:43Z   2025-04-04T09:02:43Z   1
operator.operator-infra.shell                                              job-listener                       OK        Listening on /var/run/cilium/shell/sockets/shell.sock           2025-04-04T09:02:43Z   2025-04-04T09:02:43Z   1
operator.operator-infra.shell                                              job-shell-0                        Stopped   Running                                                         2025-04-04T09:03:24Z   2025-04-04T09:03:24Z   1
operator.operator-infra.shell                                              job-shell-1                        Stopped   Running                                                         2025-04-04T09:03:51Z   2025-04-04T09:03:51Z   1
operator.operator-infra.shell                                              job-shell-10                       Stopped   Running                                                         2025-04-04T09:17:04Z   2025-04-04T09:17:04Z   1
operator.operator-infra.shell                                              job-shell-11                       OK        Running                                                         2025-04-04T09:47:19Z   2025-04-04T09:47:19Z   1
operator.operator-infra.shell                                              job-shell-2                        Stopped   Running                                                         2025-04-04T09:03:53Z   2025-04-04T09:03:53Z   1
operator.operator-infra.shell                                              job-shell-3                        Stopped   Running                                                         2025-04-04T09:04:53Z   2025-04-04T09:04:53Z   1
operator.operator-infra.shell                                              job-shell-4                        Stopped   Running                                                         2025-04-04T09:05:28Z   2025-04-04T09:05:28Z   1
operator.operator-infra.shell                                              job-shell-5                        Stopped   Running                                                         2025-04-04T09:05:29Z   2025-04-04T09:05:29Z   1
operator.operator-infra.shell                                              job-shell-6                        Stopped   Running                                                         2025-04-04T09:05:45Z   2025-04-04T09:05:45Z   1
operator.operator-infra.shell                                              job-shell-7                        Stopped   Running                                                         2025-04-04T09:05:59Z   2025-04-04T09:05:59Z   1
operator.operator-infra.shell                                              job-shell-8                        Stopped   Running                                                         2025-04-04T09:06:16Z   2025-04-04T09:06:16Z   1
operator.operator-infra.shell                                              job-shell-9                        Stopped   Running                                                         2025-04-04T09:13:36Z   2025-04-04T09:13:36Z   1
```

To stop this clutter, this commit changes the logic to close
the health scope when a connection is closed.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

joamaki commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -309,6 +309,10 @@ spec:
         {{- end }}
+        # accessing debug shell socket from k8s debug containers
+        - name: shell-sockets
+          mountPath: /var/run/cilium/shell/sockets
```
Would it make sense to just have `/var/run/cilium/sockets` that'd have both `shell.sock` and whatever envoy puts there?

mhofstetter replied:
I thought about this, but envoys sockets are shared via hostmount volumemount (because these UNIX domain sockets need to shared between Pods (agent <-> envoy (in its own daemonset)). IMO it's better to handle the shell sockets separately - also to avoid confusion if agent & operator run on the same k8s node.

mhofstetter replied:
current implementation no longer uses volumemounts

joamaki commented on contrib/testing/debug-mount-shell.yaml:
```diff
@@ -0,0 +1,9 @@
+# Can be used to connect to the Cilium debug shell from a k8s debugging container
+# e.g. for the Cilium Operator
+#
+# kubectl -n kube-system debug cilium-operator-7dcff8d46c-79sp6 -it --image=localhost:5000/cilium/cilium-dev:local --custom=contrib/testing/debug-mount-shell.yaml --profile=general -- bash
```
Could we instead of exporting the volume mount etc. just have `cilium-operator shell` command that'd do something like `cilium-dbg/cmd/shell.go`? This is what I'd eventually like to happen for the agent as well (e.g. cilium-dbg binary goes away, you only have `cilium-agent` that has a `shell` sub-command).

mhofstetter replied:
haven't thought about this. has it's pros and cons - easier to use - but not as easy to pipe into other linux tooling (except one does it on the client)

mhofstetter replied:
Refactored this so that `cilium-agent` & `cilium-operator` binaries have the `shell` subcommand

joamaki commented on pkg/shell/shell.go:
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: Apache-2.0
```
Hmm I do wonder if this should move into `cilium/hive` instead. I don't see a reason why not to tightly couple it into it. Could provide the socket path via `hive.Options` and the hive cobra commands (`command.go`) could have a `shell` sub-command implementing basically what is currently in `cilium-dbg/cmd/shell.go`. Then `cilium-dbg shell` would just exec `cilium-agent hive shell`.

mhofstetter replied:
Honestly - I was surprised to see that `cilium/hive` already provides built-in shell commands. Even though it makes sense to interactively play with the hive cell - i wonder whether we should add even more "Cilium related" "runtime" aspects to this repo - as it's actual "just" provides a dependency injection framework :thinking: Personally I'd keep it in the `cilium/cilium` repo for the time being.

mhofstetter replied:
as discussed offline - let's keep this in `cilium/cilium` for the time being.

mhofstetter committed:
shell: extract shell cobra command

This commit extracts the shell client cobra command into
`pkg/shell/client` as a preparation to reuse the command
in the cilium agent and operator.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
agent/operator: add shell command to Cilium Agent & Operator

This commit adds the `shell` command (shell client) to the
Cilium Agent & Operator root command.

This way, one can connect to the cilium debug shell in containers
where `cilium-dbg` doesn't exist.

e.g. for the Cilium Operator

```
kubectl -n kube-system exec -it cilium-operator-86649d6bc8-sfp9v -- cilium-operator-generic shell
```

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
shell: create runtime directory if it doesn't exist

This commit checks if the runtime directory needs to be created
before the shell listener tries to create the socket.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

tommyp1ckles reviewed (approved):
Nice, this will be very useful.
